add trainer hook functions

gcramer23 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60785 add trainer hook functions**
* #60784 refactor ps benchmark

This pr adds hook functions for the trainers.

Differential Revision: [D29697299](https://our.internmc.facebook.com/intern/diff/D29697299)

gcramer23 committed:
add trainer hook functions

[ghstack-poisoned]

mrshenli reviewed (commented):
Any reason these hooks need to live in separate files? Can they just live in on `hooks.py`?

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

gcramer23 commented:
> Any reason these hooks need to live in separate files? Can they just live in on `hooks.py`?

I updated the code. Added all hooks to hooks.py inside trainer directory.

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainer/hooks.py:
```diff
@@ -0,0 +1,96 @@
+        state (object): maintains state during the training process
+        bucket (GradBucket): gradient bucket
+    """
+    cref = state.cref
```
(Apologize if I have asked this question before) What does "c" in "cref" stands for?

gcramer23 replied:
It stands for class.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainer/hooks.py:
```diff
@@ -0,0 +1,96 @@
+
+def hybrid_hook(state, bucket):
+    r"""
+    A ddp communication hook that uses two process groups.
```
this is gloo for sparse and nccl for dense, right?

gcramer23 replied:
Yes, I will update the comment.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainer/hooks.py:
```diff
@@ -0,0 +1,96 @@
+        cref.record_start("hook_c10d_metric", key, "gloo_sparse_allreduce")
+        tensor = tensor.coalesce()
+        tensor = tensor / state.process_group.size()
+        c10d.all_reduce(tensor, op=c10d.ReduceOp.SUM)
```
(no need to change anything) curious, any reason that we choose to initialize the default process group with gloo, and create a non-default nccl one in the state, instead of creating two non-default ones?

gcramer23 replied:
No reason, just a different implementation. Only thing is the default process group async returns a work handle, not a torch future.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainer/hooks.py:
```diff
@@ -0,0 +1,96 @@
+        tensor = tensor / state.process_group.size()
+        c10d.all_reduce(tensor, op=c10d.ReduceOp.SUM)
+        cref.record_end("hook_c10d_metric", key)
+        fut = torch.futures.Future()
```
(can be done in followup PRs) I believe gloo's work handle now supports `get_future`. Using that might give us better perf.

gcramer23 replied:
Yes, I would need to use a non-default process group.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainer/hooks.py:
```diff
@@ -0,0 +1,96 @@
+        fut = state.process_group.allreduce(tensor).get_future()
+
+        def callback(fut):
+            cref.record_end("hook_future_metric", key)
```
I assume this is using CUDAMetrics?

gcramer23 replied:
Yes.

mrshenli reviewed (approved):
LGTM. Left some minor comments.

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

gcramer23 committed:
Update on "add trainer hook functions"


This pr adds hook functions for the trainers.

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29697299).
